File: disjoint_set/quick_find/disjoint_set.py
class Disjoint:
    def __init__(self, size):
        self.roots = []
        for i in range(size):
            self.roots.append(i)
    def __str__(self):
        s = "["
        for i in range(len(self.roots)):
            if i == len(self.roots) - 1:
                s += str(self.roots[i])
            else:
                s += str(self.roots[i]) + ", "
        s += "]"
        return s
    def find(self, x):
        return self.roots[x]
    def unionSet(self,x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            for i in range(len(self.roots)):
                if self.roots[i] == rootY:
                    self.roots[i] = rootX
    def connected(self,x,y):
        return self.find(x) == self.find(y)


def main():
    ds = Disjoint(10)
    ds.unionSet(1, 2);
    ds.unionSet(2, 5);
    ds.unionSet(5, 6);
    ds.unionSet(6, 7);
    ds.unionSet(3, 8);
    ds.unionSet(8, 9);
    print(ds.connected(1, 5))
    print(ds.connected(5, 7))
    print(ds.connected(4, 9))

    print(ds)

if __name__ == "__main__":
    main()

File: disjoint_set/quick_find/disjoint_set.cpp
#include <iostream>
#include <vector>

using namespace std;

class Disjoint {
  public:
  /*
    @Instantiates the disjoint set
    Each node's root node is itself during initial construction
   */
  Disjoint(int size) : roots(size) {
    for(int i=0; i<size; i++) {
      roots[i] = i;
    }
  }
  /*
    @Returns the root of a given node in the set
  */
  int find(int x) {
    return roots[x];
  }
  /*
    @Unions two vertices and makes their root nodes the same
   */
  void unionSet(int x, int y) {
    int rootX = find(x);
    int rootY = find(y);
    if(rootX != rootY) {
      for (int i = 0; i < roots.size(); i++) {
        if (roots[i] == rootY) {
          roots[i] = rootX;
        }
      }
    }
  }

  bool connected(int x, int y) {
    return find(x) == find(y);
  }

  void print() {
    for (int i = 0; i < roots.size(); i++) {
      cout << roots[i] << " ";
    }
  }

  private:
  vector<int> roots;
};


int main() {
  Disjoint ds(10);
  ds.unionSet(1, 2);
  ds.unionSet(2, 5);
  ds.unionSet(5, 6);
  ds.unionSet(6, 7);
  ds.unionSet(3, 8);
  ds.unionSet(8, 9);

  cout << ds.connected(1, 5) << endl;
  cout << ds.connected(5, 7) << endl;
  cout << ds.connected(4, 9) << endl;

  ds.unionSet(9, 4);
  cout << ds.connected(4, 9) << endl;

  ds.print();

  cout << "Testing C++ constructors" << endl;

  ds = Disjoint(8);
  ds.unionSet(1, 2);
  ds.unionSet(2, 5);
  ds.unionSet(5, 6);
  ds.unionSet(6, 7);
  ds.unionSet(3, 8);
  ds.unionSet(8, 9);

  cout << ds.connected(1, 5) << endl;
  cout << ds.connected(5, 7) << endl;
  cout << ds.connected(4, 9) << endl;

  ds.unionSet(9, 4);
  cout << ds.connected(4, 9) << endl;
  cout << endl;
  ds.print();
  return 0;
}
